container/lxd/image.go

```diff
@@ -126,41 +131,65 @@ func (s *Server) FindImage(
 // CopyRemoteImage accepts an image sourced from a remote server and copies it
 // to the local cache
 func (s *Server) CopyRemoteImage(
-	sourced SourcedImage, aliases []string, callback environs.StatusCallbackFunc,
+	ctx context.Context, sourced SourcedImage, aliases []string, callback environs.StatusCallbackFunc,
 ) error {
 	logger.Debugf("Copying image from remote server")
 
 	newAliases := make([]api.ImageAlias, len(aliases))
 	for i, a := range aliases {
 		newAliases[i] = api.ImageAlias{Name: a}
 	}
-
 	req := &lxd.ImageCopyArgs{Aliases: newAliases}
-	op, err := s.CopyImage(sourced.LXDServer, *sourced.Image, req)
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-	// Report progress via callback if supplied.
-	if callback != nil {
-		progress := func(op api.Operation) {
-			if op.Metadata == nil {
+	progress := func(op api.Operation) {
+		if op.Metadata == nil {
+			return
+		}
+		for _, key := range []string{"fs_progress", "download_progress"} {
+			if value, ok := op.Metadata[key]; ok {
+				_ = callback(status.Provisioning, fmt.Sprintf("Retrieving image: %s", value.(string)), nil)
 				return
 			}
-			for _, key := range []string{"fs_progress", "download_progress"} {
-				if value, ok := op.Metadata[key]; ok {
-					_ = callback(status.Provisioning, fmt.Sprintf("Retrieving image: %s", value.(string)), nil)
-					return
-				}
-			}
-		}
-		_, err = op.AddHandler(progress)
-		if err != nil {
-			return errors.Trace(err)
 		}
 	}
 
-	if err := op.Wait(); err != nil {
+	var op lxd.RemoteOperation
+	attemptDownload := func() error {
+		var err error
+		op, err = s.CopyImage(sourced.LXDServer, *sourced.Image, req)
+		if err != nil {
+			return err
+		}
+		// Report progress via callback if supplied.
+		if callback != nil {
+			_, err = op.AddHandler(progress)
+			if err != nil {
+				return err
+			}
+		}
+		if err := op.Wait(); err != nil {
+			return err
+		}
+		return nil
+	}
+	err := retry.Call(retry.CallArgs{
+		Clock:       s.clock,
+		Attempts:    3,
+		Delay:       15 * time.Second,
```

Comment:
Why is the delay so big? Especially considering the backoff is double delay?

How did you arrive at 3 for an attempt count?

Reply:
So we want to back off with a long delay because trying it in quick succession doesn't fix it, but with a larger delay should help. Personally I would add more retries, but we should log out in debug output when we do so.

Reply:
We need some commentary about this in the code as it's unexpected.

Reply:
That's fair. To be honest, this config is pretty arbitrary, other than the exponential backoff, since I believe this is standard practice for retry strategies for web requests

3 seemed like a reasonable number of retried, and given how long it can take for the lxd client to fail during a download, 15s seemed reasonable.

These feelings are not strongly held though and I'm happy to basically anything requested